pkg/fqdn/cache.go

```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
+	d.Lock()
+	defer d.Unlock()
+
+	entry, exists := d.deletes[ip.String()]
+	if !exists {
+		return
+	}
+	entry.ActiveAt = now
+}
+
+// SetCTGCTime marks the start of the most recent CT GC. This must be set after
+// all MarkActive calls complete to avoid races. When now is later than an
+// active timestamp set with MarkActive the entry is no longer active. Thus,
+// this call acts as a gating function for what data is returned by GC.
+func (d *DNSPendingDeletes) SetCTGCTime(now time.Time) {
+	d.lastCTGCUpdate = now
+}
+
+// ForceExpire is used to clear entries irrespective of their active status (or
+// them being updated at all). Only entries with DeletePendingAt times before
+// expireLookupBefore are considered. Each name in an entry is matched against
+// nameMatcher (nil is match all) and when an entry no longer has any valid
+// names will it be removed outright.
+// Note that all parameters must match, if provided. `time.Time{}` is
+// considered not-provided for time parameters.
+// expireLookupsBefore requires an entry to have been enqueued before the
+// specified time in order to remove it.
+// nameMatch will remove any DNS names that match.
+func (d *DNSPendingDeletes) ForceExpire(expireLookupsBefore time.Time, nameMatch *regexp.Regexp) (namesAffected []string) {
+	d.Lock()
+	defer d.Unlock()
+
+	var toDelete []*DNSPendingDelete
+
+	for _, entry := range d.deletes {
+		// Do not expire entries that were enqueued after expireLookupsBefore, but
+		// only if the value is non-zero
+		if !expireLookupsBefore.Equal(time.Time{}) && entry.DeletePendingAt.After(expireLookupsBefore) {
+			continue
+		}
+
+		// An entry has multiple names, collect the ones that should remain (i.e.
+		// do not match nameMatch)
+		var newNames []string
+		for _, name := range entry.Names {
+			if nameMatch != nil && !nameMatch.MatchString(name) {
+				newNames = append(newNames, name)
+			} else {
+				namesAffected = append(namesAffected, name)
+			}
+		}
+		entry.Names = newNames
+
+		// Delete the entry outright if no names remain
+		if len(entry.Names) == 0 {
+			toDelete = append(toDelete, entry)
+		}
+	}
+
+	// Delete the entries that are now empty
+	for _, entry := range toDelete {
+		delete(d.deletes, entry.IP.String())
+	}
+
+	return namesAffected
+}
+
+// DumpActive returns copies of still-active delete entries
+func (d *DNSPendingDeletes) DumpActive() (active []*DNSPendingDelete) {
+	d.Lock()
+	defer d.Unlock()
+
+	for _, del := range d.deletes {
+		if d.canDelete(del) {
+			continue
+		}
+
+		cpy := &DNSPendingDelete{
+			IP:              del.IP,
+			DeletePendingAt: del.DeletePendingAt,
+			ActiveAt:        del.ActiveAt,
+		}
+		cpy.Names = append(cpy.Names, del.Names...)
+
+		active = append(active, cpy)
+	}
+
+	return active
+}
+
+// MarshalJSON encodes DNSPendingDeletes into JSON. Only the DNSPendingDelete
+// entries are encoded.
+func (d *DNSPendingDeletes) MarshalJSON() ([]byte, error) {
+	aux := struct {
```

Comment:
add comment here stating we are doing this to avoid making `Deletes` as a public field for `DNSPendingDeletes`